### DISCUSSION TITLE
Working on same file in different branches, keeping branches seperated and clean

### DISCUSSION THREAD

**Discussion by theMeanRat** [community/community#21765] [New to GitHub]
<p>Dear all</p>
<p>I am working on a large scale project that has been messed up by our predecessors. We are planning to clean all **bleep**ty code (maybe by rebuilding it in Laravel), but before that we have to develop the current application further. Otherwise our customer comes to a halt in production.</p>
<p>One developer thought of a way to develop each new issue in a new branch. We have 3 main branches: “development” - “acceptation” - “master”. When a new issue arises, we branch off of master to a new branch called “issueX”. Another issue? Make branch off of master “issueY”.</p>
<p>But when you work on both issues yourself, and you have changes on the same file code gets intermixed. Say you have made changes to “module.php” within branch issueX. After that you checkout to branch issueY and make other changes to again “module.php”. After committing, branch issueY will contain some code of issueX.</p>
<p>Another problem is with hotfixes. I really have to change hotfix code in GitLab with the online IDE editor, because otherwise I definately will commit some development code to the hotfix branch, and therefor be putting some code live in production after deploying the hotfix!</p>
<p>How can I checkout to different branches, committing to them without committing code from other branches? I need to be able to get my environment synced with the branch 100%. So even when I have created a new file for issueX, after checking out to issueY that new file must be gone!</p>
<p><strong>Probable solution</strong></p>
<p>Right now I am thinking of these 2 commands:</p>
<h1>git checkout issueY</h1>
<h1>git reset --hard</h1>
<p>Would that be a good way to go? Or is that a weird strategy? I would love to have your thoughts about this. Thanks a whole lot in advance!</p>
<p>BG</p>
<p>Maarten</p>

**Comment by airtower-luna** [comment]
<p>“git checkout BRANCH” will already set files to the state they are on that branch, as long as they are “clean” (that is, have no uncommited changes), so you can skip the reset. It will not remove untracked files, but that shouldn’t be a problem as long as you don’t <em>add</em> them on a branch they don’t belong to.</p>
<p>The main challenge here seems to be to branch from the right source branch, e.g. to branch both issueX and issueY from master, not issueY from issueX, because otherwise you’ll indeed end up with commits for X in the Y branch.</p>

**Comment by theMeanRat** [comment]
<p>Below an example. In this case, although, I used issueA and issueB instead of X/Y. But the problem doesnt change of course.</p>
<p><img src="https://user-images.githubusercontent.com/47529424/181066783-7d24cf19-21d4-4fec-8767-717b5141f493.png" alt="gitquestion_issueX.png" data-base62-sha1="zw5kixqLZRwyxAgyBf66Z6d0Ha8" width="430" height="195"></p>
<p>When checking out to another branch I also thought that files would change to that branch. But see above: this is issueA. I just added this line. Committed it and pushed it to that branch:</p>
<h1>git push origin issueA</h1>
<p><img src="https://user-images.githubusercontent.com/47529424/181066785-e182f2b4-448f-437e-b679-b205871d1831.png" alt="gitquestion_issueB.png" data-base62-sha1="paQFsu5mpO5f6obBhIKQWbFB85l" width="520" height="173"></p>
<p>This is branch issueB. The line is not there yet. So now I want to work on issueB.</p>
<h1>git checkout -b issueB</h1>
<h1>git branch</h1>
<p><img src="https://user-images.githubusercontent.com/47529424/181066787-8511fe4f-add2-41e0-90ae-83704faa8852.png" alt="gitquestion_whichBranch.png" data-base62-sha1="8QVh01SkYk0gEJil99r9wLBaeZa" width="274" height="118"></p>
<p>So I am on the right branch</p>
<h1>less index.php (the file I just changed)</h1>
<p><img src="https://user-images.githubusercontent.com/47529424/181066789-b28dda4a-8dce-4418-a1d1-7b96416930c4.png" alt="gitquestion_stillIssueX.png" data-base62-sha1="b7D0fKkoTaBLUrGGDBPEuhf5EEM" width="367" height="212"></p>
<p>Even when I do <strong># git log</strong> I see the commit added to issueA! So why is it mingling with other branches here? Is it because I use  <strong>origin</strong> when pushing to branch  <strong>issueA </strong> in one of the first steps?</p>
<p>Thanks.</p>

**Reply by Dharmil-Shah3** [reply]
`git stash` might be helpful to you.

[Reference link](https://www.atlassian.com/git/tutorials/saving-changes/git-stash)

suppose you working on branchA,
save the work of branchA in stash.
i.e) `git stash save "work of branchA"`

Now, you can go to other branch and work on them.
You can also stash the work of that branch.

Now, come back to branchA and see that stash of branchA is at what position in stash using following command
`git stash list`

And reapply all changes from stash to the current working copy of files using **pop** OR **apply** with stash command.
`git stash pop stash@{1}`
here the **stash@{1}** says that the stash you want to pop is at index **1** in stash

**Comment by airtower-luna** [comment, marked as answer]
<p>It’s because you didn’t change the branch between your work on issueA (including the push) and creating the new branch. If you are on the issueA branch and do</p>
<p>$ git checkout -b issueB</p>
<p>the issueB branch will start at the latest commit on the issueA branch. That’s the problem.</p>
<p>You have two ways to solve that (examples assuming you want to start from master):</p>
<ol>
<li>Change to the right branch before creating the new one:<br>
$ git checkout master<br>
$ git checkout -b issueB</li>
<li>Use the optional “start point” argument to checkout when creating the branch:<br>
$ git checkout -b issueB master</li>
</ol>
<p>Either way your new branch “issueB” will start at the current state of “master”. You can use a commit ID instead of a branch name if you want.</p>

**Comment by theMeanRat** [comment]
<p>Thanks so much for enlighting me. This is much appreceated and very valuable to me.</p>
<p>Best regard,<br>
Maarten</p>
